go/marshal/encode.go

```diff
@@ -223,18 +260,29 @@ func structEncoder(t reflect.Type, parentStructTypes []reflect.Type) encoderFunc
 
 	parentStructTypes = append(parentStructTypes, t)
 	fields, structType, originalFieldIndex := typeFields(t, parentStructTypes)
-	if structType != nil {
+
+	// It won't be possible to use the pre-computed struct type if it implements
+	// MarshalOptionsGetter, because the struct name can change at runtime.
+	isOptsGetter := t.Implements(optionsGetterType)
```

Comment:
This code appears to have changed a lot, but it hasn't. The reason I moved around the conditional blocks: it's easier to follow now with the extra check for `isOptsGetter` (IMO).